### ISSUE TITLE
KLJSLoss

### ISSUE THREAD

**Issue by Zhang-O** [PaddlePaddle/PaddleOCR#6747]
KLJSLoss is not KL divergence, neither JS divergence.
Is this a customed loss func which is not a existed func used in mathematics?

```python
class KLJSLoss(object):
    def __init__(self, mode='kl'):
        assert mode in ['kl', 'js', 'KL', 'JS'
                        ], "mode can only be one of ['kl', 'js', 'KL', 'JS']"
        self.mode = mode

    def __call__(self, p1, p2, reduction="mean"):

        loss = paddle.multiply(p2, paddle.log((p2 + 1e-5) / (p1 + 1e-5) + 1e-5))

        if self.mode.lower() == "js":
            loss += paddle.multiply(
                p1, paddle.log((p1 + 1e-5) / (p2 + 1e-5) + 1e-5))
            loss *= 0.5
        if reduction == "mean":
            loss = paddle.mean(loss, axis=[1, 2])
        elif reduction == "none" or reduction is None:
            return loss
        else:
            loss = paddle.sum(loss, axis=[1, 2])

        return loss
```

**Comment by Zhang-O** [comment]
Another question following above:
DML computes KL divergence using two maps of size (B,C,H,W)  from two students, 
but from codes below, we can see that it just uses softmax in W direction, this seems weird.
In general usages, softmax could change a vector to a probability distribution which always occured in the last layer of classification tasks.
Here, softmax is just applied in one direction of feature map. I guess it should be used all over the feature map to get the distribution. Or just use the MSE to get the difference of two feature maps.

```python3
self.act = nn.Softmax(axis=-1)
```

```python3
class DMLLoss(nn.Layer):
    """
    DMLLoss
    """

    def __init__(self, act=None, use_log=False):
        super().__init__()
        if act is not None:
            assert act in ["softmax", "sigmoid"]
        if act == "softmax":
            self.act = nn.Softmax(axis=-1)
        elif act == "sigmoid":
            self.act = nn.Sigmoid()
        else:
            self.act = None

        self.use_log = use_log
        self.jskl_loss = KLJSLoss(mode="js")

    def _kldiv(self, x, target):
        eps = 1.0e-10
        loss = target * (paddle.log(target + eps) - x)
        # batch mean loss
        loss = paddle.sum(loss) / loss.shape[0]
        return loss

    def forward(self, out1, out2):
        if self.act is not None:
            out1 = self.act(out1) + 1e-10
            out2 = self.act(out2) + 1e-10
        if self.use_log:
            # for recognition distillation, log is needed for feature map
            log_out1 = paddle.log(out1)
            log_out2 = paddle.log(out2)
            loss = (
                self._kldiv(log_out1, out2) + self._kldiv(log_out2, out1)) / 2.0
        else:
            # for detection distillation log is not needed
            loss = self.jskl_loss(out1, out2)
        return loss
```

**Comment by pmspku** [comment]
> Another question following above: DML computes KL divergence using two maps of size (B,C,H,W) from two students, but from codes below, we can see that it just uses softmax in W direction, this seems weird. In general usages, softmax could change a vector to a probability distribution which always occured in the last layer of classification tasks. Here, softmax is just applied in one direction of feature map. I guess it should be used all over the feature map to get the distribution. Or just use the MSE to get the difference of two feature maps.
> 
> ```python
> self.act = nn.Softmax(axis=-1)
> ```
> 
> ```python
> class DMLLoss(nn.Layer):
>     """
>     DMLLoss
>     """
> 
>     def __init__(self, act=None, use_log=False):
>         super().__init__()
>         if act is not None:
>             assert act in ["softmax", "sigmoid"]
>         if act == "softmax":
>             self.act = nn.Softmax(axis=-1)
>         elif act == "sigmoid":
>             self.act = nn.Sigmoid()
>         else:
>             self.act = None
> 
>         self.use_log = use_log
>         self.jskl_loss = KLJSLoss(mode="js")
> 
>     def _kldiv(self, x, target):
>         eps = 1.0e-10
>         loss = target * (paddle.log(target + eps) - x)
>         # batch mean loss
>         loss = paddle.sum(loss) / loss.shape[0]
>         return loss
> 
>     def forward(self, out1, out2):
>         if self.act is not None:
>             out1 = self.act(out1) + 1e-10
>             out2 = self.act(out2) + 1e-10
>         if self.use_log:
>             # for recognition distillation, log is needed for feature map
>             log_out1 = paddle.log(out1)
>             log_out2 = paddle.log(out2)
>             loss = (
>                 self._kldiv(log_out1, out2) + self._kldiv(log_out2, out1)) / 2.0
>         else:
>             # for detection distillation log is not needed
>             loss = self.jskl_loss(out1, out2)
>         return loss
> ```

this is Rdrop

**Comment by LDOUBLEV** [comment]
@Zhang-O 
- It is really inappropriate to use js naming, I will pull a request to correct it later.
- The reason why softmax is only calculated in the last dimension is that the features of the recognition model to calculate DMLloss are already two-dimensional.

**Comment by Zhang-O** [comment]
@LDOUBLEV  
Thanks for your reply!
I could understand  that CRNN-DMLLoss applies softmax to last dim of features which shape is  (batch-size, max_len, classes).
The softmax's result is  a probability distribution of char classes, which is a natural arguement of KL divergence function.

But in the context of DBNet DML-distillation,  the shape of feature map is （batch-size, channels, height, width）.
Applying softmax to last dim of features means getting a probability distribution from width direction. 
And what does that mean?

From my viewpoint, configs located in "configs/det/ch_PP-OCRv3/ch_PP-OCRv3_det_dml.yml",
the key "act" should change to sigmoid in the context of det.

```yml
Loss:
  name: CombinedLoss
  loss_config_list:
  - DistillationDMLLoss:
      model_name_pairs:
      - ["Student", "Student2"]
      maps_name: "thrink_maps"
      weight: 1.0
      act: "softmax"
      model_name_pairs: ["Student", "Student2"]
      key: maps
```

**Comment by Zhang-O** [comment]
Thank you for your fix !!!
